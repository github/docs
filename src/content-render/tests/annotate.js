import { describe, expect, test } from 'vitest'
import cheerio from 'cheerio'

import { renderContent } from '#src/content-render/index.js'

const example = `
\`\`\`yaml annotate
# The name of the workflow as it will appear in the "Actions" tab of the GitHub repository.
name: Post welcome comment

# Add the \`pull_request\` event, so that the workflow runs automatically
# every time a pull request is created.
on:
  pull_request:
    types: [opened]
\`\`\`
`

describe('annotate', () => {
  test('renders annotations', async () => {
    // We don't normally use snapshots,
    // but in this case its a short and concise example
    // that won't change regularly.
    // If it fails, study the output and make sure it's correct.
    // If it is indeed correct, run `vitest --updateSnapshot` to update it.
    expect(await renderContent(example)).toMatchSnapshot()
  })

  test('renders bash with hash bang annotations', async () => {
    const example = `
\`\`\`bash annotate
# The next line is the hash bang
#!/usr/bin/env bash

# Sample comment
echo "Hello, world!"
\`\`\`
`.trim()
    const res = await renderContent(example)
    const $ = cheerio.load(res)

    const headerCode = $('header pre').text()
    expect(headerCode).toMatch(example.split('\n').slice(1, -1).join('\n'))
    const rows = $('.annotate-row')
    const notes = $('.annotate-note p', rows)
    const noteTexts = notes.map((i, el) => $(el).text()).get()
    expect(noteTexts).toEqual(['The next line is the hash bang', 'Sample comment'])
    const codes = $('.annotate-code pre', rows)
    const codeTexts = codes.map((i, el) => $(el).text()).get()
    expect(codeTexts).toEqual(['#!/usr/bin/env bash', 'echo "Hello, world!"'])
  })

  test("doesn't complain if the first comment is empty", async () => {
    const example = `
\`\`\`yaml annotate copy
#
name: Create and publish a Docker image

# Configures this workflow to run every time...
on:
  push:
    branches: ['release']

\`\`\`
`.trim()
    const res = await renderContent(example)
    const $ = cheerio.load(res)

    const headerCode = $('header pre').text()
    expect(headerCode).toMatch(example.split('\n').slice(1, -1).join('\n'))
    const rows = $('.annotate-row')
    const notes = $('.annotate-note p', rows)
    const noteTexts = notes.map((i, el) => $(el).text()).get()
    expect(noteTexts).toEqual(['Configures this workflow to run every time...'])
    const codes = $('.annotate-code pre', rows)
    const codeTexts = codes.map((i, el) => $(el).text()).get()
    expect(codeTexts).toEqual([
      'name: Create and publish a Docker image',
      "on:\n  push:\n    branches: ['release']",
    ])
  })
})
