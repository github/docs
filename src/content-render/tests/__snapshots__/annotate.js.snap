// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`annotate > renders annotations 1`] = `
"<div class="annotate beside"><div class="annotate-header"><header class="d-flex flex-items-center flex-justify-between p-2 text-small rounded-top-1 border-top border-left border-right"><span class="flex-1">YAML</span><div class="annotate-toggle"><button name="annotate-display" value="beside" type="button" class="annotate-option">Beside</button><button name="annotate-display" value="inline" type="button" class="annotate-option">Inline</button></div><button class="js-btn-copy btn btn-sm tooltipped tooltipped-nw" aria-label="Copy YAML code to clipboard" data-clipboard="1746955726"><svg version="1.1" width="16" height="16" viewBox="0 0 16 16" class="octicon octicon-copy" aria-hidden="true"><path d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z"></path><path d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z"></path></svg></button><pre hidden data-clipboard="1746955726"># The name of the workflow as it will appear in the "Actions" tab of the GitHub repository.
name: Post welcome comment

# Add the \`pull_request\` event, so that the workflow runs automatically
# every time a pull request is created.
on:
  pull_request:
    types: [opened]
</pre></header></div><div class="annotate-beside"><div class="annotate-row"><div class="annotate-code"><pre><code class="hljs language-yaml"><span class="hljs-attr">name:</span> <span class="hljs-string">Post</span> <span class="hljs-string">welcome</span> <span class="hljs-string">comment</span></code></pre></div><div class="annotate-note"><p>The name of the workflow as it will appear in the "Actions" tab of the GitHub repository.</p></div></div><div class="annotate-row"><div class="annotate-code"><pre><code class="hljs language-yaml"><span class="hljs-attr">on:</span>
  <span class="hljs-attr">pull_request:</span>
    <span class="hljs-attr">types:</span> [<span class="hljs-string">opened</span>]</code></pre></div><div class="annotate-note"><p>Add the <code>pull_request</code> event, so that the workflow runs automatically
every time a pull request is created.</p></div></div></div><div class="annotate-inline"><pre><code class="hljs language-yaml"><span class="hljs-comment"># The name of the workflow as it will appear in the "Actions" tab of the GitHub repository.</span>
<span class="hljs-attr">name:</span> <span class="hljs-string">Post</span> <span class="hljs-string">welcome</span> <span class="hljs-string">comment</span>

<span class="hljs-comment"># Add the \`pull_request\` event, so that the workflow runs automatically</span>
<span class="hljs-comment"># every time a pull request is created.</span>
<span class="hljs-attr">on:</span>
  <span class="hljs-attr">pull_request:</span>
    <span class="hljs-attr">types:</span> [<span class="hljs-string">opened</span>]
</code></pre></div></div>"
`;
